Sources/Plasma/FeatureLib/pfPython/cyMiscGlue2.cpp

```diff
@@ -40,34 +40,143 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 
 *==LICENSE==*/
 
+#include "cyMisc.h"
+
 #include <Python.h>
-#include "pyKey.h"
 
-#include "cyMisc.h"
-#include "pyGlueHelpers.h"
+#include "pyEnum.h"
 #include "pyColor.h"
+#include "pyGlueHelpers.h"
+#include "pyKey.h"
 #include "pyPlayer.h"
-#include "pyEnum.h"
+#include "plPythonCallable.h"
 
-// for enums
+#include "plMessage/plConfirmationMsg.h"
 #include "plNetCommon/plNetCommon.h"
 #include "plResMgr/plLocalization.h"
 #include "plMessage/plLOSRequestMsg.h"
 
+namespace plPythonCallable
+{
+    static inline void IBuildTupleArg(PyObject* tuple, size_t idx, plConfirmationMsg::Result value)
+    {
+        PyTuple_SET_ITEM(tuple, idx, PyLong_FromSsize_t((Py_ssize_t)value));
+    }
+};
 
-PYTHON_GLOBAL_METHOD_DEFINITION(PtYesNoDialog, args, "Params: selfkey,dialogMessage\nThis will display a Yes/No dialog to the user with the text dialogMessage\n"
-            "This dialog _has_ to be answered by the user.\n"
-            "And their answer will be returned in a Notify message.")
+PYTHON_GLOBAL_METHOD_DEFINITION_WKEY(PtYesNoDialog, args, kwargs,
+            "Params: cb, message, /, dialogType\n"
+            "This will display a confirmation dialog to the user with the text `message` "
+            "This dialog _has_ to be answered by the user, "
+            "and their answer will be returned in a Notify message or callback given by `cb`.")
 {
-    PyObject* keyObj = nullptr;
+    const char* keywords[]{ "", "", "dialogType", nullptr };
+    constexpr std::string_view kErrorMsg = "PtYesNoDialog expects a ptKey or callable, "
+                                           "a string or localization path, and an optional int.";
+    PyObject* cbObj;
     ST::string text;
-    if (!PyArg_ParseTuple(args, "OO&", &keyObj, PyUnicode_STStringConverter, &text) || !pyKey::Check(keyObj)) {
-        PyErr_SetString(PyExc_TypeError, "PtYesNoDialog expects a ptKey and a string");
+    plConfirmationMsg::Type dialogType = plConfirmationMsg::Type::YesNo;
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs,
+                                     "OO&|I", const_cast<char**>(keywords),
+                                     &cbObj,
+                                     PyUnicode_STStringConverter, &text,
+                                     &dialogType)) {
+        PyErr_SetString(PyExc_TypeError, kErrorMsg.data());
         PYTHON_RETURN_ERROR;
     }
-    pyKey* key = pyKey::ConvertFrom(keyObj);
-    cyMisc::YesNoDialog(*key, text);
-    PYTHON_RETURN_NONE
+
+    plConfirmationMsg::Callback cb;
+    if (pyKey::Check(cbObj)) {
+        cb = pyKey::ConvertFrom(cbObj)->getKey();
+    } else if (PyCallable_Check(cbObj)) {
+        cb = plPythonCallable::BuildCallback<plConfirmationMsg::Result>("PtYesNoDialog", cbObj);
```

Comment:
@zrax I'm a little unhappy that this function requires manual specification of the callback arguments. I'm sure there's a way to erase those, but I'm drawing a blank. Any ideas?

Reply:
I don't think you can in this case, at least not without moving the construction of the tuple outside the callback...

Reply:
I got it down to specifying the variant index, which is good enough... I guess...